converts EuiForm (and family) to TypeScript

dimitropoulos opened the pull request:
closes: https://github.com/elastic/eui/issues/2892

kibanamachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

dimitropoulos commented on src/components/basic_table/basic_table.tsx:
```diff
@@ -16,7 +16,7 @@ import { CommonProps } from '../common';
 import { get } from '../../services/objects';
 import { EuiFlexGroup, EuiFlexItem } from '../flex';
-import { EuiCheckbox } from '../form/checkbox/checkbox';
+import { EuiCheckbox } from '../form';
```
from what I can tell the desired way of using these in the project is to import from `/form` directly rather than `/form/thing/other_thing.tsx` so I updated only for the sake of cleaning it up (well, that, and also there are `// @ts-ignore`'s that are no longer necessary which I also cleaned up wherever I found them.  If I assumed correctly, please let me know and I'll happily revert the import styles.

chandlerprall replied:
_Depends_. We want to export (from the top-level _index.ts_ files) only pieces that should be useful/used outside of EUI. In many of the changes made, the smaller building block components became exported (e.g. EuiRangeHighlight, EuiRangeTrack, EuiRangeWrapper).

dimitropoulos replied:
Gotcha.  I wasn't sure since before all of the ones you mentioned were exported from their respective subdirectory ([for example](https://github.com/elastic/eui/blob/688b7792d9d44cf3d4a73ce2aab8f5d06a4160d4/src/components/form/range/index.ts#L1)), but then [not exported directly from form](https://github.com/elastic/eui/blob/master/src/components/form/index.js).  The problem I had was that it didn't feel consistent which ones are exported and not.

To phrase the question another way: there's no reason I can see (other than perhaps reducing number of import lines elsewhere in the codebase?) to export, for example, `EuiRangeHighlight` from `src/components/form/range/index.ts` but _not_ from `src/components/form/index.ts` since anything that needs `EuiRangeHighlight` that's internal to the codebase can just get it directly from `src/components/form/range/range_highlight.tsx`.  Does that make sense?

In other words - and I hate to do this to you - if there isn't a rule I can follow to know which things should be exported and which things should not be could you possibly either make a commit where you include or remove the ones you think should be exposed?

I can always reference [`src/components/form/index.ts`](https://github.com/elastic/eui/blob/master/src/components/form/index.ts) from before this PR and just include the things that were there - will that work?

chandlerprall replied:
> In other words - and I hate to do this to you - if there isn't a rule I can follow to know which things should be exported and which things should not be could you possibly either make a commit where you include or remove the ones you think should be exposed?
>
> I can always reference src/components/form/index.ts from before this PR and just include the things that were there - will that work?

No worries, go ahead and follow the previous exports from _src/components/form/index.ts_ - we can circle back around later with more, but at least that will keep things As They Are. I'm starting to work on some much better (and needed) documentation around these things and will try to come up with a rule that can be followed in the future.

dimitropoulos replied:
no problemo! I totally get it.  updated in https://github.com/elastic/eui/pull/2896/commits/1711df6056baeefa4d76f58ecf3167939edb41c7 to match the previous exports exactly.

1. I will just mention one case where you might consider allowing.  There are 5 components that take an `options` prop that's an array of some derived type, for example:  [`EuiCheckboxGroupOption`](https://github.com/dimitropoulos/eui/blob/1711df6056baeefa4d76f58ecf3167939edb41c7/src/components/form/checkbox/checkbox_group.tsx#L13), [`EuiRadioGroupOption`](https://github.com/dimitropoulos/eui/blob/1711df6056baeefa4d76f58ecf3167939edb41c7/src/components/form/radio/radio_group.tsx#L13), [`EuiSuperSelectOption`](https://github.com/dimitropoulos/eui/blob/1711df6056baeefa4d76f58ecf3167939edb41c7/src/components/form/super_select/super_select_control.tsx#L19) (especially this one, since it's generic), [`EuiSelectOptionProps`](https://github.com/dimitropoulos/eui/blob/1711df6056baeefa4d76f58ecf3167939edb41c7/src/components/form/select/select.tsx#L16) (also, now would be a good time, perhaps, to rename this to `EuiSelectOption` - which seems to be appropriate when compared to the others in this list).

    The reason for exporting the above is that it's a little complicated for Eui consumers since otherwise you have to extract the value of the options array using a helper which many projects (including this one) don't have `ValueOf<EuiCheckboxGroupProps['options']>`.  For ComboBox, for example, we exported the comparable type (i.e. of the `options` prop) since it's quite a bit more convenient for consumers of Eui.

2. There's also `EuiSwitchEvent` since otherwise the user will have to extract the argument to get the type and it's feasible that if someone is wrapping this callback with a classmember that they may want the type - although addmitedly they could always do `classMember: EuiSwitchProps['onChange'] = evnet => {}`.

dimitropoulos replied:
Since I can't really respond during my workday - I went ahead and pushed to another branch, [`dimitropoulos/eui/form-ts-options`](https://github.com/dimitropoulos/eui/commits/form-ts-options) where I made the following three commits in following with the above, should you be interested in adding in the types I mentioned:

1. https://github.com/dimitropoulos/eui/commit/6249fd2bbf92427b3d25daa3cc53beddc4055915 adds in the 4 options types mentioned above
2. https://github.com/dimitropoulos/eui/commit/e5258e2ddf50d9c1991fed2c5b1cb1eb158e0ac8 renames `EuiSelectOptionProps` to `EuiSelectOption` to match all other `options` props named in the project. (this was committed separately so that you can take it or leave it separately)
3. https://github.com/dimitropoulos/eui/commit/09a1e299c26e7174736f315b8e5b26b034d68451 exports `EuiSwitchEvent`

Again, really, please feel free to cherrypick in these commits _or not_.  I'm happy with whatever you think is best: just trying to keep the end users in mind :smile:.

chandlerprall replied:
Perfect, and thank you for wrangling all of those! I've cherrry-picked each of those 3 commits and pushed to your branch. Running some final tests and then I'll merge this in!

dimitropoulos commented on src/components/color_picker/color_stops/color_stop_thumb.tsx:
```diff
@@ -328,10 +330,7 @@ export const EuiColorStopThumb: FunctionComponent<EuiColorStopThumbProps> = ({
           </EuiFlexItem>
           <EuiFlexItem grow={false}>
-            <EuiFormRow
-              grow="false"
```
it appears that `EuiFormRow` does not implement `grow`, but please verify

dimitropoulos commented on src/components/color_picker/color_stops/color_stop_thumb.tsx:
```diff
@@ -374,7 +373,6 @@ export const EuiColorStopThumb: FunctionComponent<EuiColorStopThumbProps> = ({
                 <EuiFormRow
                   label={hexLabel}
                   display="rowCompressed"
-                  readOnly={readOnly}
```
it appears that `EuiFormRow` does not implement `readOnly`, but please verify

dimitropoulos commented on src/components/datagrid/column_selector.tsx:
```diff
@@ -150,9 +147,10 @@ export const useColumnSelector = (
-                            }: React.FormEvent<HTMLInputElement>) => {
+                            onChange={(event: EuiSwitchEvent) => {
+                              const {
+                                target: { checked },
```
changing from `currentTarget` to `target` is a behavior change brought about by the types (initially) but [when I dug deeper](https://github.com/elastic/eui/blob/master/src/components/form/switch/switch.tsx#L60) it appears that actually `target` is correct.  please confirm :)

chandlerprall replied:
`target` and `currentTarget` should always be the same element, in this specific instance. But I think `target` is going to be more future proof.

dimitropoulos commented on src/components/form/form.tsx:
```diff
@@ -1,20 +1,33 @@
       <ul>
         {errorTexts.map(error => (
-          <li className="euiForm__error" key={error}>
+          <li className="euiForm__error" key={makeId()}>
```
from my quick debugging, `error` is, indeed, a Node, and thus not satisfactory for a `key`.  The choices are either to become heretics and use the index as a key (which, really is fine in some cases when you *know* the order isn't going to change - and perhaps this is one of those cases), or we can generate a shortId.  I looked around the project for other exemplars and it seemed like `makeId` is the most common approach used in this case so that's what I went with.  If that was wrong: please let me know how you'd like to proceed.

chandlerprall replied:
We have no qualms using the index as key

dimitropoulos replied:
cool beans :)

https://github.com/elastic/eui/pull/2896/commits/f4db3150eac71bb409d82b56c397150800eee304

dimitropoulos commented on src/components/form/form.tsx:
```diff
@@ -42,17 +55,9 @@ export const EuiForm = ({ children, className, isInvalid, error, ...rest }) => {
 
   return (
-    <div className={classes} {...rest}>
+    <form className={classes} {...rest}>
```
this is a (potentially) impactful change that was originally demanded by the types, but I looked around for a while and it seems like there's a decent chance this is actually correct.  I didn't find any CSS selector problems, but such things can be tough to track down so I want to be very cautious about this change.  Any insight from someone more familiar with Eui will be useful here to not make the wrong choice.  The other option would be to loosen the types from above, which (presently, and before this PR) were calling out HTMLFormElement directly.

chandlerprall replied:
@cchaos I know our lack of actual `form` usage has come up before; you'd be much better at addressing this comment.

thompsongl replied:
#2272 discusses this, but we haven't yet come to a real answer.
This would be a breaking change, though, and something we'd want to avoid right now.

snide replied:
Yep, comments from that thread still make sense. Still think this isn't worth it for now. If we added it it should be optional as a non-default prop / new component. Messing with forms isn't something we should go into lightly and would have a lot of downstream ramifications.

dimitropoulos replied:
I figured as much.  The types now reflect reality a little better.  If nothing else, hopefully this will helps make it clear to downstream consumers that it's a `div` not a `form`.

I should make sure to say clearly, though, that this change to reflect that reality in the types is a difference from what [was here previously](https://github.com/elastic/eui/blob/688b7792d9d44cf3d4a73ce2aab8f5d06a4160d4/src/components/form/index.d.ts#L10) which had it extending `FormHTMLAttributes<HTMLFormElement>`.

This means that all of these guys:
```ts
    interface FormHTMLAttributes<T> extends HTMLAttributes<T> {
        acceptCharset?: string;
        action?: string;
        autoComplete?: string;
        encType?: string;
        method?: string;
        name?: string;
        noValidate?: boolean;
        target?: string;
    }
```

Used to (erroneously) not report errors, but now it will:
![Screenshot_20200224_193157](https://user-images.githubusercontent.com/15232461/75203541-568da880-573c-11ea-9307-71ee598cff72.png)

Since this is, then, technically a bugfix it's not (depending on your reading of semver) a breaking change, but maybe just something to be aware of.

https://github.com/elastic/eui/pull/2896/commits/c27837ee892b1e7d55ebcd5123666f4f9e587eaf

dimitropoulos commented on src/components/form/form_control_layout/form_control_layout_delimited.tsx:
```diff
@@ -12,7 +12,9 @@ import {
 } from './form_control_layout';
 
-type EuiFormControlLayoutDelimitedProps = Partial<EuiFormControlLayoutProps> & {
+export type EuiFormControlLayoutDelimitedProps = Partial<
```
I went file by file and made sure there were no props missing and there were a few (like this), that were not previously exported.  I'm assuming that wasn't on purpose, but if it was (for whatever reason) please let me know and I'll update.

chandlerprall replied:
Thanks!! We've not been the most dedicated at ensuring these are exported.

snide commented:
Thanks again for the PR. We'll mostly be silent over the weekend, but will give this a review on Monday.

chandlerprall commented on src/components/datagrid/column_selector.tsx:
```diff
@@ -11,14 +11,11 @@ import {
 import { EuiFlexGroup, EuiFlexItem } from '../flex';
-// @ts-ignore-next-line
-import { EuiSwitch, EuiFieldText } from '../form';
+import { EuiSwitch, EuiFieldText, EuiSwitchEvent } from '../form';
```
Since I called out not exporting the smaller building blocks, I want to specifically point to `EuiSwitchEvent` as something that _does appear useful_ outside of EUI, as it defines the event's modification for use by a consuming application.

i.e. please keep `EuiSwitchEvent` as a top-level export

dimitropoulos replied:
thanks for the clarification!

chandlerprall commented on src/components/form/form.tsx:
```diff
@@ -1,20 +1,33 @@
+    error?: ReactNode | ReactNode[];
+  };
+
+export const EuiForm: FC<EuiFormProps> = ({
```
Prefer the full `FunctionComponent` name

dimitropoulos replied:
ahh! yes! sorry!  didn't realize that this project has that preference.  :)

I'm gonna blame this one on muscle memory, haha.

FYI, you can codify this with the [ban-types](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/ban-types.md) rule methinks:

```
{
  "rules": {
    "@typescript-eslint/ban-types": [
      "error",
      {
        "types": {
          "React.FC": {
            "message": "Use React.FunctionComponent instead",
            "fixWith": "React.FunctionComponent"
          }
        }
      }
    ]
  }
}
```

fixed in https://github.com/elastic/eui/pull/2896/commits/28c97105026632892a9acd31b61e2e7e2589553f

chandlerprall replied:
Working on documenting these preferences, too! Lots going on :) Thank you again for your patience on all of these.

dimitropoulos replied:
most def: didn't mean it as a criticism in the slightest :smile: 

also, give yourself a break! this is a great project.  really.

chandlerprall commented on src/components/form/form_error_text/form_error_text.tsx:
```diff
@@ -1,11 +1,16 @@
-import React, { FunctionComponent, HTMLAttributes } from 'react';
+import React, { FC, HTMLAttributes } from 'react';
```
As above, prefer `FunctionComponent` name

dimitropoulos replied:
ahh! yes! sorry!  didn't realize that this project has that preference.  :)

I'm gonna blame this one on muscle memory, haha.

FYI, you can codify this with the [ban-types](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/ban-types.md) rule methinks:

```
{
  "rules": {
    "@typescript-eslint/ban-types": [
      "error",
      {
        "types": {
          "React.FC": {
            "message": "Use React.FunctionComponent instead",
            "fixWith": "React.FunctionComponent"
          }
        }
      }
    ]
  }
}
```

fixed in https://github.com/elastic/eui/pull/2896/commits/28c97105026632892a9acd31b61e2e7e2589553f

dimitropoulos commented on src/components/color_picker/color_stops/color_stop_thumb.tsx:
```diff
@@ -310,15 +313,14 @@ export const EuiColorStopThumb: FunctionComponent<EuiColorStopThumbProps> = ({
                     readOnly={readOnly}
-                    min={isRangeMin || min == null ? null : localMin}
-                    max={isRangeMax || max == null ? null : localMax}
+                    min={isRangeMin || min === undefined ? undefined : localMin}
```
Sure: no problem: changed in https://github.com/elastic/eui/pull/2896/commits/c25853f660ca77145a3fe9419d0fa08bef4c6b87

I was wondering about this but I defaulted to following the types that were already here: https://github.com/elastic/eui/blob/688b7792d9d44cf3d4a73ce2aab8f5d06a4160d4/src/components/form/field_number/field_number.tsx#L23

Would you like me to update that so that instead of `max? number` they all (i.e. including `EuiColorStopThumbProps.max`/`.min` and `EuiColorStopsProps.max`/`.min`) read `max?: number | null`?

chandlerprall replied:
No, let's keep the definition of `optional type` (e.g. `max?: number` / `max: number | undefined`. The `null`s were a far more prominent in style in non-TS code so we need to support them in a few places, but largely move away from them.

dimitropoulos commented:
fixed `form.test.tsz` -> `form.test.tsx` in https://github.com/elastic/eui/pull/2896/commits/ffc0e27873e0dd67961203e5bee028229b390d19 (whoops! :sweat_smile: I was moving very fast the night I did these.. heh.)

Otherwise, I think I've addressed all other feedback and just need some direction on what/how you want to be exported.  I know with this kind of thing (since I've now exported everything it'll be literally just deleting single lines in the various `index.ts` in subdirectories like `field_number/`) it can be easier to just make the change rather than writing up detailed descriptions of the changes in a PR.  I'm happy to do anything that works well for you all - so feel free to do what's easiest for you.

dimitropoulos commented:
@chandlerprall see https://github.com/elastic/eui/pull/2896#discussion_r384199690 and https://github.com/elastic/eui/pull/2896#discussion_r384205569, but otherwise I think we're all set here :smiley:!
